bittensor/_dataset/dataset_impl.py

```diff
@@ -1,536 +1,490 @@
-        self.__infinite_dataset_iterator = None
+        if len(self.dataset_name) == 0 or self.dataset_name == 'default':
+            self.dataset_name = self.available_datasets
+        self.datasets = self.dataset_name = self.dataset_name[:self.max_datasets]
         self.no_tokenizer = no_tokenizer
-        self.IPFS_fails = 0
-        self.backup_dataset_cap_size = 5e7 # set 50MB limit per folder
-        self.IPFS_fails_max = 10
+        self.tokenizer = None if self.no_tokenizer else bittensor.tokenizer()
+        self.data_dir =  data_dir
+        self.save_dataset = save_dataset
+        self.load_dataset = load_dataset
+        self.run_generator= run_generator
+        self.buffer_size = buffer_size
         self.num_batches = num_batches
         self.max_directories = max_directories
 
-        # Retrieve a random slice of the genesis dataset
-        self.data = []
-        self.data_reserved = []
+        # we need to build the dataset or load existing text file hashes
+        # notice the heirarchy of ipfs hashes is DATASET -> FOLDER -> TEXT HASH, 
+        # we want to flatten each dataset FOLDER -> TEXT HASH into FOLDER*TEXT
+    
 
-        # Used to refresh corpus if we've exhausted the whole dataset
-        self.refresh_corpus = True
+        self.build_datasets(datasets=self.dataset_name, load=self.load_dataset, save=self.save_dataset)
+
+        self.data_queue = Queue(buffer_size)
+        # this runs the a thread that has its own asyncio loop. 
+        # The loop is passed into nested async functions to use loo.run_until_complete function
+        if self.run_generator:
+            # the thread manager is used for running a background thread
+            self.thread_manager = ThreadManager()
+            # start the genrator
+            self.thread_manager.submit(fn=self.sample_generator, kwargs=dict(queue=self.data_queue, loop=asyncio.new_event_loop()))
+    def sample_generator(self, 
+                         queue:Queue, 
+                         loop:'asyncio.loop'=None, 
+                         return_json:bool=False):
+
+        """ Sample generator on seperate thread with its own asyncio loop for generating
+            background samples while the user fetches them in the foreground.
+        Args:
+            queue (Queue):
+                Queue for feeding the samples through for __getitem__ to fetch.
+            batch_size (int):
+                Batch size of the samples.
+            sequence_length (int):
+                Sequence Length of the samples.
+            loop:'asyncio.loop'=None, 
+                        return_json:bool=False
+        """
+        
+        # this is for starting a new thread
+        # the loop needs to be set within the new thread
+        if loop != None:
+            asyncio.set_event_loop(loop)
 
-        self.build_hash_table()
+        # chunk the text hashes into batch_sie chunks
+        text_hash_batch_list = self.chunk(self.all_text_file_metas,
```

Comment:
I don't see this variable used in this function. do you mean to iterate over it below?